File: Sources/MIDI2CI/ProcessInquirySession.swift
import Foundation
import MIDI2

/// Minimal Process Inquiry session supporting capInquiry/capReply,
/// messageReport/messageReportReply, and endReport.
public final class ProcessInquirySession {
    /// Reported capabilities/filters (key -> level)
    public var filters: [String: UInt8]
    /// Device ID scope for replies (0x00-0x0F, 0x7E, 0x7F)
    public var deviceId: UInt8

    public init(filters: [String: UInt8] = [:], deviceId: UInt8 = 0x7F) {
        self.filters = filters
        self.deviceId = deviceId
    }

    private func isValidDeviceId(_ id: UInt8) -> Bool {
        id <= 0x0F || id == 0x7E || id == 0x7F
    }

    /// Clamp requested filters to what this session supports (min of requested/support) and drop unsupported keys.
    private func clampFilters(_ requested: [String: UInt8]?) -> [String: UInt8] {
        guard let requested else { return [:] }
        var reply: [String: UInt8] = [:]
        for (key, reqVal) in requested {
            guard let supportedVal = filters[key], supportedVal > 0 else { continue }
            reply[key] = min(reqVal, supportedVal)
        }
        return reply
    }

    /// Handle a Process Inquiry body and return zero or one reply.
    public func handle(_ body: MidiCiProcessInquiryBody) -> MidiCiProcessInquiryBody? {
        // Validate Device ID scope (M2-101-UM Table 40)
        guard isValidDeviceId(deviceId) else { return nil }
        switch body.command {
        case .capInquiry:
            return MidiCiProcessInquiryBody(command: .capReply, filters: filters)
        case .messageReport:
            // Acknowledge start by echoing the intersection of requested and supported filters
            let clamped = clampFilters(body.filters)
            return MidiCiProcessInquiryBody(command: .messageReportReply, filters: clamped)
        case .endReport:
            // No payload expected in reply; end of reporting sequence acknowledged by no response
            return nil
        default:
            return nil
        }
    }
}

File: midi2.js/src/process-inquiry-session.ts
import { ProcessInquiryEvent } from "./types";

function isValidDeviceId(id: number): boolean {
  return Number.isInteger(id) && (id <= 0x0f || id === 0x7e || id === 0x7f);
}

function clampFilters(requested: Record<string, number> | undefined, supported: Record<string, number>): Record<string, number> {
  if (!requested) return {};
  const out: Record<string, number> = {};
  for (const [key, reqVal] of Object.entries(requested)) {
    if (!Number.isInteger(reqVal) || reqVal < 0 || reqVal > 0x7f) continue;
    if (key === "messageDataControl" && reqVal !== 0x00 && reqVal !== 0x01 && reqVal !== 0x7f) continue;
    const sup = supported[key];
    if (sup && sup > 0) {
      out[key] = Math.min(reqVal, sup);
    }
  }
  return out;
}

/**
 * Minimal Process Inquiry runtime helper: echoes capability (capReply) and clamps messageReport filters.
 */
export class ProcessInquirySession {
  private readonly supported: Record<string, number>;
  private readonly deviceId: number;

  constructor(opts: { filters?: Record<string, number>; deviceId?: number } = {}) {
    this.supported = opts.filters ?? {};
    this.deviceId = opts.deviceId ?? 0x7f;
  }

  handle(evt: ProcessInquiryEvent): ProcessInquiryEvent | null {
    if (!isValidDeviceId(this.deviceId)) return null;
    switch (evt.command) {
      case "capInquiry":
        return { kind: "processInquiry", group: evt.group, command: "capReply", filters: this.supported };
      case "messageReport": {
        const clamped = clampFilters(evt.filters, this.supported);
        return { kind: "processInquiry", group: evt.group, command: "messageReportReply", filters: clamped };
      }
      case "endReport":
        return null;
      default:
        return null;
    }
  }
}
